### ISSUE TITLE
Clustering Issue (same center point)

### ISSUE THREAD

**Issue by Farah189** [masqm/Faster-Mean-Shift#1]
Hi, I was trying to use your code for effectively using GPU machine for mean shift clustering. It is always giving me just one centre point, irrespective of the number of clusters. Maybe I am not using the command correctly. Following is a data for one such point that make two clusters; one around 6 x-coordinate and the other around 47 x-coordinate. It is giving one cluster centre at [[ 31.30769231 110.53846154]]. I was using the following command:

cluster  = MeanShiftCosine(bandwidth=3, cluster_all=True, GPU=true).fit(c)
center_points = cluster.cluster_centers_.astype('int')

My data is:
       [[  4 112], [  5 111], [  5 112], [  6 111], [  6 112], [  6 113], [  7 111], [  7 112], [  7 113], [  8 112],
       [ 45 109], [ 45 110], [ 46 108], [ 46 109], [ 46 110], [ 46 111], [ 47 108], [ 47 109], [ 47 110], [ 47 111], [ 48 109], [ 48 110],
       [ 48 111], [ 49 109], [ 49 110], [ 49 111]]

Also, the clustering time is the same as was used by the CPU.

**Comment by masqm** [comment]
Hi Farah189,

I'm very happy to see someone use our code. LOL

Well, if I'm not wrong, I think your simulation data is not related based on cosine similarity. However, the distance metric in our clustering algorithm is based on cosine similarity. Thus, the data will be cluster based on the included angle between to vectors.

Here is a link to briefly introduce cosine similarity: https://www.machinelearningplus.com/nlp/cosine-similarity/#:~:text=Cosine%20similarity%20is%20a%20metric,in%20a%20multi%2Ddimensional%20space.

Therefore, if you want to use our algorithm to cluster your data, you need to change the bandwidth to a very small value. (Rough estimate: 0.05 or 0.1) 

And for the time complexity issue,  we are quite surprised the clustering time is the same as the CPU. Actually, we think our algorithm should slower than CPU. Because the amount of data in your simulation is very small,  it should be more time-consuming by using GPU. According to the experiment in our paper, the points should be at least 10k to have a clear advantage. We recommend using make_blobs() in sklearn to generate these points.

In the end, if you have any questions, please contact us again. And, please cite our paper, if you would like to use our code in your research.

Best
Masqm

**Comment by Farah189** [comment]
You are absolutely right. My simulation data is kind of blobs in 2D plane. I shared the smallest possible data to support my question. I reduced the bandwidth to a smaller value too but it didn't help. I read the main part of your paper after posting a comment here. I was already using MeanShift (MS) from sklearn.clusters but it is consuming major time of my simulation so I started looking for GPU based MS implementation. I should have read the paper properly before bugging you. You can also specify it in the reading section of github to avoid such queries in the future. And thanks for the clarification here too.

**Comment by masqm** [comment]
Hi! 

Sorry for taking so long. Recently, I used this code again and found there is a serious error in the bandwidth update, which cause the problems that you asked. I just fixed it and update the code. Please try this new code again, if you still need to use it. Thanks!

Best,
Masqm
